File: packages/design-ir/src/query.ts
import type { DesignDocument, DesignNode, NodeSelector } from "./types";

function hasBinding(node: DesignNode, key: string, expected: string): boolean {
  const direct = node[key];
  if (typeof direct === "string") return direct === expected;
  const semantic = node.semantic?.[key];
  return typeof semantic === "string" && semantic === expected;
}

function descendantOf(document: DesignDocument, node: DesignNode, ancestorId: string): boolean {
  let current = node.parent_id;
  while (current !== null) {
    if (current === ancestorId) return true;
    current = document.nodes[current]?.parent_id ?? null;
  }
  return false;
}

export function queryNodes(document: DesignDocument, selector: NodeSelector): readonly DesignNode[] {
  if (selector.id) {
    const node = document.nodes[selector.id];
    return node ? [node] : [];
  }
  return Object.values(document.nodes).filter((node) => {
    if (selector.role !== undefined && node.role !== selector.role) return false;
    if (selector.kind !== undefined && node.kind !== selector.kind) return false;
    if (selector.parent_id !== undefined && node.parent_id !== selector.parent_id) return false;
    if (selector.locked !== undefined && Boolean(node.locked) !== selector.locked) return false;
    if (
      selector.brand_binding !== undefined &&
      !hasBinding(node, "brand_binding", selector.brand_binding)
    ) return false;
    if (
      selector.asset_binding !== undefined &&
      !hasBinding(node, "asset_id", selector.asset_binding)
    ) return false;
    if (
      selector.frame_id !== undefined &&
      node.id !== selector.frame_id &&
      !descendantOf(document, node, selector.frame_id)
    ) return false;
    return true;
  });
}

File: packages-py/design_ir/query.py
from __future__ import annotations

from .models import DesignDocument


def _binding(node: dict, key: str) -> object:
    if key in node:
        return node[key]
    semantic = node.get("semantic")
    return semantic.get(key) if isinstance(semantic, dict) else None


def _descendant_of(document: DesignDocument, node: dict, ancestor_id: str) -> bool:
    current = node.get("parent_id")
    while current is not None:
        if current == ancestor_id:
            return True
        parent = document["nodes"].get(current)
        current = parent.get("parent_id") if isinstance(parent, dict) else None
    return False


def query_nodes(document: DesignDocument, selector: dict) -> list[dict]:
    node_id = selector.get("id")
    if node_id is not None:
        node = document["nodes"].get(node_id)
        return [node] if isinstance(node, dict) else []

    result: list[dict] = []
    for node in document["nodes"].values():
        if not isinstance(node, dict):
            continue
        if "role" in selector and node.get("role") != selector["role"]:
            continue
        if "kind" in selector and node.get("kind") != selector["kind"]:
            continue
        if "parent_id" in selector and node.get("parent_id") != selector["parent_id"]:
            continue
        if "locked" in selector and bool(node.get("locked")) != selector["locked"]:
            continue
        if "brand_binding" in selector:
            if _binding(node, "brand_binding") != selector["brand_binding"]:
                continue
        if "asset_binding" in selector and _binding(node, "asset_id") != selector["asset_binding"]:
            continue
        if "frame_id" in selector:
            frame_id = selector["frame_id"]
            if node.get("id") != frame_id and not _descendant_of(document, node, frame_id):
                continue
        result.append(node)
    return result
